sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java

```diff
@@ -393,89 +394,58 @@ public double getProgress() {
+      // DoFn.processElement() call and access it later on. We are still
+      // calling next() here, causing an update to byteCount. If user's
+      // code really accesses more element's pieces later on, their byte
+      // count would accrue against a future element. This is not ideal,
+      // but still approximately correct.
+      if (getIsLazy()) {
+        advance();
       }
     }
 
-    private static final int RESERVOIR_SIZE = 10;
-    private static final int SAMPLING_THRESHOLD = 30;
-    private long samplingToken = 0;
-    private long nextSamplingToken = 0;
-    private Random randomGenerator = new Random();
+    // Lowest sampling probability: 0.001%.
+    private static final int SAMPLING_TOKEN_UPPER_BOUND = 1000000;
+    private static final int SAMPLING_CUTOFF = 10;
+    private int samplingToken = 0;
+    private Random randomGenerator = ThreadLocalRandom.current();
 
     private boolean shouldSampleElement() {
       // Sampling probability decreases as the element count is increasing.
-      // We unconditionally sample the first samplingCutoff elements. Calculating
-      // nextInt(samplingToken) for each element is expensive, so after a threshold, calculate the
-      // gap to next sample.
-      // https://erikerlandson.github.io/blog/2015/11/20/very-fast-reservoir-sampling/
-
-      // Reset samplingToken if it's going to exceed the max value.
-      if (samplingToken + 1 == Long.MAX_VALUE) {
-        samplingToken = 0;
-        nextSamplingToken = getNextSamplingToken(samplingToken);
-      }
-
-      samplingToken++;
-      // Use traditional sampling until the threshold of 30
-      if (nextSamplingToken == 0) {
-        if (samplingToken <= RESERVOIR_SIZE
-            || randomGenerator.nextInt((int) samplingToken) < RESERVOIR_SIZE) {
-          if (samplingToken > SAMPLING_THRESHOLD) {
-            nextSamplingToken = getNextSamplingToken(samplingToken);
-          }
-          return true;
-        }
-      } else if (samplingToken >= nextSamplingToken) {
-        nextSamplingToken = getNextSamplingToken(samplingToken);
-        return true;
-      }
-      return false;
-    }
-
-    private long getNextSamplingToken(long samplingToken) {
-      double gap =
-          Math.log(1.0 - randomGenerator.nextDouble())
-              / Math.log(1.0 - RESERVOIR_SIZE / (double) samplingToken);
-      return samplingToken + (int) gap;
+      // We unconditionally sample the first samplingCutoff elements. For the
+      // next samplingCutoff elements, the sampling probability drops from 100%
+      // to 50%. The probability of sampling the Nth element is:
+      // min(1, samplingCutoff / N), with an additional lower bound of
+      // samplingCutoff / samplingTokenUpperBound. This algorithm may be refined
+      // later.
+      samplingToken = Math.min(samplingToken + 1, SAMPLING_TOKEN_UPPER_BOUND);
+      return randomGenerator.nextInt(samplingToken) < SAMPLING_CUTOFF;
```

Comment:
Even if we switch to a thread-local RNG, I'm concerned that calling nextInt on every single value will be more expensive than doing the fast reservoir sampling algorithm that we had here before.

Reply:
Yeah this is still WIP and marked it as so. I'm still trying to root cause why the existing implementation shows up on FlexWordCount profiles.